42/README.md

```diff
@@ -0,0 +1,71 @@
+```
+shortname: [42/HARD-FORKS]
+name: Handling Forks
+type: information
+Status: raw
+editor: Vanshdeep Singh <vanshdeep@bigchaindb.com>
+```
+
+## Description
+
+Currently, there is no active strategy to address breaking changes in BigchainDB 2.0 i.e. there is no general approach to migration such that breaking changes in transaction specs can be handled. This document is an attempt in quantifying scenarios wherein a breaking might occur and propose a generalized strategy to handle these situations.
+
+## Hard fork scenarios
+1. Change in any of the transaction specs
+    - Change in signature algorithm
+    - Change in hashing functions
+    - Change in schema
+2. Change in storage schema
+
+## Handling Hard Forks
+
+### Type 1 scenarios
+In case of change to spec it might not be possible to port all existing validated transaction to the new format. In such cases one needs to make a distinction between transactions which have already been validated and stored in the Blockchain from new incoming transactions. Furthermore, a suitable trigger is needed so that the network can collectively jump to the new logic else there could be situations wherein the network might not be able to commit any new blocks because of honest nodes running different validation logic. So, it is apparent that a mechanism for transitioning to this new validation logic is needed. Below is one such procedure to achieve the same,
```

Comment:
My thinking about the "Type 1 scenario" was that if a new transaction spec (e.g. v3) gets implemented, then nodes can accept new v2 and v3 transactions for a while. Before a node upgrades its BigchainDB Server software to a version that recognizes v3 transactions, it will just think that v3 transactions are invalid, so for a while, v3 transactions will be considered invalid by the network. Once > 2/3 of the nodes upgrade, then v3 transactions will have a chance of being considered valid. No TEP or vote is needed.

Later, the nodes could upgrade their BigchainDB Server software to software that rejects _new_ v2 transactions, but can still validate old v2 transactions.

This process can be repeated for v4, v5, etc.

Reply:
@ttmc we need to distinguish old transactions from new transactions. Upon replay, how does a validator know that he needs to switch to `v2` at height `5` and then to `v3` at height `20`? The approach described here basically leverages the election process to address these problems.

Reply:
A transaction contains a field/key "version" that says which transaction spec version it follows. Currently, the only allowed value is "2.0". The node will validate the transaction using the validation rules for that transaction version number.

Reply:
I was thinking about handling changes to the transaction spec and realized that there are at least two cases:

1a) There are no problems with the old transactions, but a new transaction spec has been devised and it is better in some ways, so people would like to use transactions following the new spec. This case can be handled as I described in my comment above, without any need for proposals, elections or changing old blocks.

1b) Some problem is found with old transactions, such as a security flaw. The course of action will depend on the nature of the problem with the old transactions, but it will have to be more drastic. For sure, new transactions following the flawed spec shouldn't be allowed, and so the proposal & election process can be used to trigger a change to a new/fixed spec. It might be necessary to do more, e.g. to prevent unspent outputs from being transferred from old/broken outputs to thieves. It all depends on what the problem is.

Reply:
> This case can be handled as I described in my comment above, without any need for proposals, elections or changing old blocks.

Yes, you are right. Also, during replay the versions of the transactions are known so we do not strictly need the election mechanism for the case when the new version can not validate old transactions.

On the other hand, the election process can make the migration smoother. Without it, how do you know if you can post new transactions already? You can run trial and error but it is costly.

Reply:
@ttmc The case I am tried to address here is 1b i.e. we might not need to do anything special in case there is new transaction spec which facilitates a new behavior (or a backwards compatible change). The main concern this strategy addresses is when the validation rules for an existing transaction have to be upgraded such that the new rules are incompatible with the older rules. The reason for doing so might be a security flaw of something else. So the general aim to ensure that no new transactions that invalid according to new validation rules are committed to the blockchain. But we also need to ensure that the older transaction if and when replayed are indeed validated. So it seems that we have an explicit declaration which indicates the activation of new rules across the network.

> It might be necessary to do more, e.g. to prevent unspent outputs from being transferred from old/broken outputs to thieves. It all depends on what the problem is.

True, depending on the scenario there would be other required changes but these changes would be a part of the new validation logic.

Reply:
It seems that Tendermint is very strict: you should never have a time interval (sequence of blocks) where some validators think a transaction is valid and others think it's invalid: they'll consider each other to be Byzantine validators! That must be avoided, so an election (TEP) will be needed to get the validators to agree on a future block height where new rules will come into effect.